File: mpp-proiect-java-culbec/MotorcycleContest/Server/src/main/java/Service/ServiceImpl.java
package Service;

import CommonUtils.IService;
import CommonUtils.Observer;
import Exceptions.AppException;
import Exceptions.ServiceException;
import Model.Participant;
import Model.Race;
import Model.User;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class ServiceImpl implements IService {
    private final Logger logger = LogManager.getLogger(this.getClass());
    private final ServiceUser serviceUser;
    private final ServiceParticipant serviceParticipant;
    private final ServiceRace serviceRace;
    private final Map<String, Observer> loggedClients = new ConcurrentHashMap<>();

    private static final int DEFAULT_THREADS_NO = 3;

    public ServiceImpl(ServiceUser serviceUser, ServiceParticipant serviceParticipant, ServiceRace serviceRace) {
        logger.traceEntry("Initializing the Service...");
        this.serviceUser = serviceUser;
        this.serviceParticipant = serviceParticipant;
        this.serviceRace = serviceRace;
        logger.traceExit("Initialized the Service!");
    }

    @Override
    public Iterable<Participant> findParticipantsByTeam(String team) throws ServiceException {
        logger.traceEntry("Finding the participants from team: {}", team);

        try {
            List<Participant> participants = (List<Participant>) this.serviceParticipant.findParticipantsByTeam(team);
            logger.info("Found {} of team: {}", String.valueOf(participants.size()), team);

            return participants;
        } catch (ServiceException e) {
            logger.error("Couldn't find the participants by team: {}", e.getMessage());
            throw new ServiceException("Couldn't find the participants by team: " + e.getMessage());
        }
    }

    @Override
    public void addParticipant(String firstName, String lastName, String team, int engineCapacity) throws ServiceException {
        logger.traceEntry("Trying to add the participant with fields {}, {}, {}, {}", firstName, lastName, team, engineCapacity);

        try {
            Participant participant = this.serviceParticipant.addParticipant(firstName, lastName, team, engineCapacity);
            logger.info("Successfully added the participant!");

            logger.info("Notifying the clients...");
            this.notifyParticipantAdded(participant);

            logger.traceExit("Notified all the clients!");
        } catch (ServiceException e) {
            logger.error("Couldn't add the participant: {}", e.getMessage());
            throw new ServiceException("Couldn't add the participant: " + e.getMessage());
        }
    }

    /**
     * Notifies the logged in clients that a new participant has been added.
     *
     * @param participant Participant that was added.
     */
    private void notifyParticipantAdded(Participant participant) {
        try (ExecutorService executor = Executors.newFixedThreadPool(DEFAULT_THREADS_NO)) {
            this.loggedClients.values().forEach(client -> executor.execute(() -> {
                try {
                    client.participantAdded(participant);
                } catch (AppException e) {
                    logger.error("Couldn't notify the client: {}", e.getMessage());
                }
            }));
        }
    }

    @Override
    public Iterable<Race> findAllRaces() throws ServiceException {
        logger.traceEntry("Finding all the races...");

        try {
            Iterable<Race> races = this.serviceRace.findAllRaces();
            logger.traceExit("Found all the races! Returning...");

            return races;
        } catch (ServiceException e) {
            logger.error("Couldn't find all the races: {}", e.getMessage());
            throw new ServiceException("Couldn't all the races: " + e.getMessage());
        }
    }

    @Override
    public Iterable<Integer> findAllRaceEngineCapacities() throws ServiceException {
        logger.traceEntry("Finding all the engine capacities...");

        try {
            Iterable<Integer> engineCapacities = this.serviceRace.findAllRaceEngineCapacities();
            logger.traceExit("Found all the engine capacities! Returning...");

            return engineCapacities;
        } catch (ServiceException e) {
            logger.error("Couldn't find all the engine capacities: {}", e.getMessage());
            throw new ServiceException("Couldn't find all the engine capacities: " + e.getMessage());
        }
    }

    @Override
    public synchronized User login(String username, String password, Observer client) throws ServiceException {
        logger.traceEntry("Trying to login the user {}", username);

        logger.info("Verifying if the user is already logged in...");
        if (loggedClients.get(username) != null) {
            logger.error("The client is already logged in!");
            throw new ServiceException("The client is already logged in!");
        }

        try {
            logger.info("Trying to retrieve the user by the passed credentials...");
            User user = this.serviceUser.login(username, password);

            logger.info("User found. Updating the logged in clients...");
            loggedClients.put(username, client);

            logger.traceExit("Returning the user...");
            return user;
        } catch (ServiceException e) {
            logger.error("Couldn't login the user: {}", e.getMessage());
            throw new ServiceException("Couldn't login the user: " + e.getMessage());
        }
    }

    @Override
    public synchronized void logout(User user, Observer client) throws ServiceException {
        logger.traceEntry("Logging out the user: {}", user);

        Observer localClient = this.loggedClients.remove(user.getUsername());
        if (localClient == null) {
            logger.error("The user is not logged in!");
            throw new ServiceException("The user is not logged in!");
        }

        logger.traceExit("The user was logged out.");
    }
}

File: mpp-proiect-csharp-culbec/Server/Service/ServiceImpl.cs
using System.Collections.Concurrent;
using Common.CommonUtils;
using Common.Exceptions;
using Model;
using log4net;

namespace Server.Service;

public class ServiceImpl : IService
{
    private static readonly ILog Log = LogManager.GetLogger(typeof(ServiceImpl));
    private readonly IServiceParticipant _serviceParticipant;
    private readonly IServiceUser _serviceUser;
    private readonly IServiceRace _serviceRace;
    private readonly ConcurrentDictionary<string, IObserver> _loggedClients = new();

    public ServiceImpl(IServiceParticipant serviceParticipant, IServiceUser serviceUser, IServiceRace serviceRace)
    {
        _serviceParticipant = serviceParticipant;
        _serviceUser = serviceUser;
        _serviceRace = serviceRace;
    }

    public IEnumerable<Participant> FindParticipantsByTeam(string team)
    {
        Log.InfoFormat("Finding the participants by team: {0}", team);

        try
        {
            var participants = _serviceParticipant.FindParticipantsByTeam(team).ToList();
            Log.InfoFormat("Found {0} participants", participants.Count);

            return participants;
        }
        catch (Exception e)
        {
            Log.ErrorFormat("Could not find the participants by team: {0}", team);
            throw new ServiceException("Could not find the participants by team: " + e.Message);
        }
    }

    public void AddParticipant(string firstName, string lastName, string team, int engineCapacity)
    {
        Log.InfoFormat("Trying to add the participants with fields: {0}, {1}, {2}, {3}", firstName, lastName, team,
            engineCapacity);

        try
        {
            var participant = _serviceParticipant.AddParticipant(firstName, lastName, team, engineCapacity);
            Log.InfoFormat("Participant added: {0}", participant);
            NotifyParticipantAdded(participant);
        }
        catch (Exception e)
        {
            Log.ErrorFormat("Could not add the participant: {0}", e.Message);
            throw new ServiceException("Could not add the participant: " + e.Message);
        }
    }

    public IEnumerable<Race> FindAllRaces()
    {
        Log.Info("Finding all the races...");

        try
        {
            var races = this._serviceRace.FindAllRaces().ToList();
            Log.InfoFormat("Found {0} races", races.Count);

            return races;
        }
        catch (Exception e)
        {
            Log.ErrorFormat("Could not find all the races: {0}", e.Message);
            throw new ServiceException("Could not find all the races: " + e.Message);
        }
    }

    public IEnumerable<int> FindAllRaceEngineCapacities()
    {
        Log.InfoFormat("Finding all the race engine capacities...");

        try
        {
            var engineCapacities = _serviceRace.FindAllRaceEngineCapacities().ToList();
            Log.InfoFormat("Found {0} engine capacities", engineCapacities.Count);

            return engineCapacities;
        }
        catch (Exception e)
        {
            Log.ErrorFormat("Could not find all the race engine capacities: {0}", e.Message);
            throw new ServiceException("Could not find all the race engine capacities: " + e.Message);
        }

    }

    public User Login(string username, string password, IObserver client)
    {
        Log.InfoFormat("Trying to login the user: {0}", username);

        Log.Info("Verifying if the user is already logged in...");
        if (_loggedClients.ContainsKey(username))
        {
            Log.Error("The user is already logged in!");
            throw new ServiceException("The user is already logged in!");
        }

        try
        {
            Log.Info("Trying to login the user...");
            var user = _serviceUser.Login(username, password);

            Log.Info("User logged in successfully.");
            _loggedClients[username] = client;

            return user;
        }
        catch (Exception e)
        {
            Log.ErrorFormat("Could not login the user: {0}", e.Message);
            throw new ServiceException("Could not login the user: " + e.Message);
        }
    }

    public void Logout(User user, IObserver client)
    {
        Log.InfoFormat("Logging out the user: {0}", user);

        _loggedClients.TryRemove(user.Username!, out var localClient);

        if (localClient == null)
        {
            Log.Error("The user is not logged in!");
            throw new ServiceException("The user is not logged in!");
        }

        Log.Info("The user was logged out.");
    }

    private void NotifyParticipantAdded(Participant participant)
    {
        foreach (var client in _loggedClients.Values)
        {
            Task.Run(() => client.ParticipantAdded(participant));
        }
    }
}